src/RpcServer/RpcServer.SmartContract.cs

```diff
@@ -141,79 +149,108 @@ private static JObject ToJson(System.Collections.Generic.IEnumerable<DataCache.T
+            return _params.Select(u => new
+            {
+                Invocation = u["invocation"]?.AsString(),
+                Verification = u["verification"]?.AsString()
+            }).Where(x => x.Invocation != null || x.Verification != null).Select(x => new Witness()
+            {
+                InvocationScript = Convert.FromBase64String(x.Invocation ?? string.Empty),
+                VerificationScript = Convert.FromBase64String(x.Verification ?? string.Empty)
+            }).ToArray();
+        }
+
         [RpcMethod]
         protected virtual JObject InvokeFunction(JArray _params)
         {
             UInt160 script_hash = UInt160.Parse(_params[0].AsString());
             string operation = _params[1].AsString();
             ContractParameter[] args = _params.Count >= 3 ? ((JArray)_params[2]).Select(p => ContractParameter.FromJson(p)).ToArray() : System.Array.Empty<ContractParameter>();
-            Signers signers = _params.Count >= 4 ? SignersFromJson((JArray)_params[3], system.Settings) : null;
+            Signer[] signers = _params.Count >= 4 ? SignersFromJson((JArray)_params[3], system.Settings) : null;
+            Witness[] witnesses = _params.Count >= 4 ? WitnessesFromJson((JArray)_params[3]) : null;
             bool useDiagnostic = _params.Count >= 5 && _params[4].GetBoolean();
 
             byte[] script;
             using (ScriptBuilder sb = new())
             {
                 script = sb.EmitDynamicCall(script_hash, operation, args).ToArray();
             }
-            return GetInvokeResult(script, signers, useDiagnostic);
+            return GetInvokeResult(script, signers, witnesses, useDiagnostic);
         }
 
         [RpcMethod]
         protected virtual JObject InvokeScript(JArray _params)
         {
             byte[] script = Convert.FromBase64String(_params[0].AsString());
-            Signers signers = _params.Count >= 2 ? SignersFromJson((JArray)_params[1], system.Settings) : null;
+            Signer[] signers = _params.Count >= 2 ? SignersFromJson((JArray)_params[1], system.Settings) : null;
+            Witness[] witnesses = _params.Count >= 2 ? WitnessesFromJson((JArray)_params[1]) : null;
             bool useDiagnostic = _params.Count >= 3 && _params[2].GetBoolean();
-            return GetInvokeResult(script, signers, useDiagnostic);
+            return GetInvokeResult(script, signers, witnesses, useDiagnostic);
+        }
+
+        [RpcMethod]
+        protected virtual JObject TraverseIterator(JArray _params)
+        {
+            Guid sid = Guid.Parse(_params[0].GetString());
+            Guid iid = Guid.Parse(_params[1].GetString());
+            int count = _params[2].GetInt32();
+            if (count > settings.MaxIteratorResultItems)
+                throw new ArgumentOutOfRangeException(nameof(count));
+            Session session;
+            lock (sessions)
+            {
+                session = sessions[sid];
+                session.ResetExpiration();
```

Comment:
If we are able to reset the expiration time, we can facilitate a DoS attack by out of memory.

Reply:
I think `session` should not be enabled on public servers. I added a setting to disable it.